File: lever/src/main/java/com/pin/lever/Action.java
package com.pin.lever;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import java.lang.ref.ReferenceQueue;
import java.lang.ref.WeakReference;

import static com.pin.lever.Lever.Priority;

abstract class Action<T> {
    static class RequestWeakReference<M> extends WeakReference<M> {
        final Action action;

        RequestWeakReference(Action action, M referent, ReferenceQueue<? super M> q) {
            super(referent, q);
            this.action = action;
        }
    }

    final Lever lever;
    final Request request;
    final WeakReference<T> target;
    final boolean noFade;
    final int memoryPolicy;
    final int networkPolicy;
    final int errorResId;
    final Drawable errorDrawable;
    final String key;
    final Object tag;

    boolean willReplay;
    boolean cancelled;

    Action(Lever lever, T target, Request request, int memoryPolicy, int networkPolicy,
           int errorResId, Drawable errorDrawable, String key, Object tag, boolean noFade) {
        this.lever = lever;
        this.request = request;
        this.target =
                target == null ? null : new RequestWeakReference<>(this, target, lever.referenceQueue);
        this.memoryPolicy = memoryPolicy;
        this.networkPolicy = networkPolicy;
        this.noFade = noFade;
        this.errorResId = errorResId;
        this.errorDrawable = errorDrawable;
        this.key = key;
        this.tag = (tag != null ? tag : this);
    }

    abstract void complete(Bitmap result, Lever.LoadedFrom from);

    abstract void error(Exception e);

    void cancel() {
        cancelled = true;
    }

    Request getRequest() {
        return request;
    }

    T getTarget() {
        return target == null ? null : target.get();
    }

    String getKey() {
        return key;
    }

    boolean isCancelled() {
        return cancelled;
    }

    boolean willReplay() {
        return willReplay;
    }

    int getMemoryPolicy() {
        return memoryPolicy;
    }

    int getNetworkPolicy() {
        return networkPolicy;
    }

    Lever getLever() {
        return lever;
    }

    Priority getPriority() {
        return request.priority;
    }

    Object getTag() {
        return tag;
    }
}

File: lever/src/main/java/com/pin/lever/Action.kt
package com.pin.lever

import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import com.pin.lever.Lever.LoadedFrom
import java.lang.ref.ReferenceQueue
import java.lang.ref.WeakReference

abstract class Action<T>(
    val lever: Lever?,
    target: T?,
    val request: Request?,
    memoryPolicy: Int,
    networkPolicy: Int,
    errorResId: Int,
    errorDrawable: Drawable?,
    key: String?,
    tag: Any?,
    noFade: Boolean
) {
    internal class RequestWeakReference<M>(
        val action: Action<*>,
        referent: M,
        q: ReferenceQueue<in M>
    ) : WeakReference<M>(referent, q)

    var target: WeakReference<T>?
    val noFade: Boolean
    val memoryPolicy: Int
    val networkPolicy: Int
    val errorResId: Int
    val errorDrawable: Drawable?
    val key: String
    val tag: Any
    var willReplay = false
    var isCancelled = false
    abstract fun complete(result: Bitmap?, from: LoadedFrom?)
    abstract fun error(e: Exception?)
    open fun cancel() {
        isCancelled = true
    }

    open fun getTarget(): T? {
        return target?.get()
    }

    fun willReplay(): Boolean {
        return willReplay
    }

    val priority: Lever.Priority
        get() = request!!.priority

    init {
        this.target = if (target == null) null else RequestWeakReference<T>(
            this,
            target,
            lever!!.referenceQueue as ReferenceQueue<in T>
        )
        this.memoryPolicy = memoryPolicy
        this.networkPolicy = networkPolicy
        this.noFade = noFade
        this.errorResId = errorResId
        this.errorDrawable = errorDrawable
        this.key = key!!
        this.tag = tag ?: this
    }
}